Add session contents for `01: Baby Steps` (#4)
unikraft/summer-of-code-2021: content/en/docs/sessions/01-baby-steps/index.md

---
title: "Session 01: Baby Steps"
linkTitle: "01. Baby Steps"
date: 2021-07-22
---
In this session we are going to understand the basic layout of the Unikraft working directory, its environment variables, as well as what the most common Unikraft specific files mean.
We are also going to take a look at how we can build basic applications and how we can extend their functionality and support by adding ported external libraries.

> ## 00. Manual kraft installation

Reviewer: ```suggestion
## 00. Manual kraft installation

```

First of all, make sure you have all the dependencies installed:
```

> apt-get install -y --no-install-recommends build-essential libncurses-dev libyaml-dev flex git wget socat bison unzip uuid-runtime;

Reviewer: ```suggestion
apt-get install -y --no-install-recommends build-essential \
        libncurses-dev libyaml-dev flex git wget socat bison \
        unzip uuid-runtime
```

By default, these are saved to ```~/.unikraft```, which is also the value of the ```UK_WORKDIR``` environment variable.
This represents the working directory for all Unikraft source code.

This is the usual layout of the Unikraft directory:
```
├── apps - This is where you would normally place existing app build
├── archs - Here we place our custom arch’s files
├── libs - This is where the build system looks for external library pool sources
├── plats - The files for our custom plats are placed here
└── unikraft - The core source code of the Unikraft Unikernel
```

There are also environment variables available for each of the above subdirectories:
```
UK_ROOT - The directory for Unikraft's core source code [default: $UK_WORKDIR/unikraft]
UK_LIBS - The directory of all the external Unikraft libraries [default: $UK_WORKDIR/libs]
UK_APPS - The directory of all the template applications [default: $UK_WORKDIR/apps]
```

After successfully running the above commands, you now have kraft installed on your system and we are ready to build some unikernels!

> ## 01. Building the Helloworld Application

Reviewer: ```suggestion
## 01. Building the Helloworld Application

```

> ### Automatic build through ```kraft```

Reviewer: ```suggestion
### Automatic build through ```kraft```

```

Makefile.uk – A Kconfig target file you can use to create compile-time toggles for your application.

build/ – All build artifacts are placed in this directory including intermediate object files and unikernel images.

.config – The selection of options for architecture, platform, libraries and your application (specified in Makefile.uk) to use with Unikraft.
```

Next, we tell kraft which platform and architecture we desire our unikernel to be built for.
```
kraft configure -p PLAT -m ARCH
```

For this example, we are going to build this application for the ```Linuxu``` (Linux Userspace) and ```KVM``` (Kernel Virtual Machine) platforms and the ```x86_64``` architecture.
Let's begin with the ```linuxu``` platform.
```
kraft configure -p linuxu -m x86_64
```

Everything is set up now, all we have left to do is tell the build system to do its magic.
```
kraft build
```

Note that this builds the final unikernel image through the stable branch of the unikraft core source code.

> If you desire to build it on another branch you should use the ```-Xv``` additional option.

Reviewer: ```suggestion
If you want to build it on another branch you should use the ```-Xv``` additional option.
```

> And that’s it! Our final unikernel binary is ready to be launched from the ```build/``` directory.

Reviewer: ```suggestion
And that's it! Our final unikernel binary is ready to be launched from the ```build/``` directory.
```

We could simply tell kraft to run it for us through the following command:
```
kraft run -p linuxu -m x86_64
```

In order to build this for the ```KVM``` platform, all we have to do is simply replace ```linuxu``` with ```kvm```.
```
kraft configure -p kvm -m x86_64
kraft build
kraft run -p kvm -m x86_64
```

Of course, this is the most basic way you can use ```kraft```, but there are many other options.
To see every option ```kraft``` has to offer, you can simply type ```kraft -h```.
If you want to know about a certain command, just follow it with the ```-h``` option.
For example, if I wanted to know more about the configure command, I would type ```kraft configure -h```.

### Manually building the ```helloworld``` application

> Let’s now learn how to build the app manually, without ```kraft```!

Reviewer: ```suggestion
Let's now learn how to build the app manually, without ```kraft```!
```

Reviewer: Say that while `kraft` is the recommended way, as it's simpler and with a nice interface, the manual approach may still be useful if you want full control over the configuration and build process. More on this in session 02: Behind the Scenes.

> First, get out of the current build’s directory and make a new one:

Reviewer: ```suggestion
First, get out of the current build's directory and make a new one:
```

```
cd ../ && mkdir 01-hello-world-manual && cd 01-hello-world-manual
```

> Now, clone the remote git repository:

Reviewer: ```suggestion
Now, clone the remote Git repository:
```

```
git clone https://github.com/unikraft/app-helloworld.git .
```

> In order to tell the build system how we want our final unikernel image to be built we can use the built-in ```ncurses menu```:

Reviewer: ```suggestion
In order to tell the build system how we want our final unikernel image to be built we can use the built-in `ncurses menu`:
```

```
make menuconfig
```

Looks like we are met with an error!
```
$ make menuconfig
Makefile:9: recipe for target 'menuconfig' failed
make: *** [menuconfig] Error 2
```

> The reason this happens is because the build system assumes we are inside ```~/.unikraft/apps/app-helloworld-manual```, which is not the case.

Reviewer: ```suggestion
The reason this happens is because the build system assumes we are inside `~/.unikraft/apps/app-helloworld-manual`, which is not the case.
```

> If you remember correctly, the build system makes use of some important environment variables, namely ```UK_WORKDIR```, ```UK_ROOT``` and ```UK_LIBS```.

Reviewer: ```suggestion
If you remember correctly, the build system makes use of some important environment variables, namely `UK_WORKDIR`, `UK_ROOT` and `UK_LIBS`.
```

> In order to properly inform the build system of our current location we will have to manually set these by prefixing whatever build command we send with the hardcoded values of where our ```Unikraft``` work directory is.

Reviewer: ```suggestion
In order to properly inform the build system of our current location we will have to manually set these by prefixing whatever build command we send with the hardcoded values of where our `Unikraft` work directory is.
```

By default, these are saved to `~/.unikraft/` directory, which is also the value of the `UK_WORKDIR` environment variable used by kraft.
This represents the working directory for all Unikraft components.
This is the usual layout of the `~/.unikraft/` directory:
```
|-- apps - This is where you would normally place existing app build
|-- archs - Here we place our custom arch's files
|-- libs - This is where the build system looks for external library pool sources
|-- plats - The files for our custom plats are placed here
`-- unikraft - The core source code of the Unikraft Unikernel
```

Apart from the general `UK_WORKDIR` environment variable that points to the overall directory, there are also environment variables available for the above subdirectories:
```
UK_ROOT - The directory for Unikraft's core source code [default: $UK_WORKDIR/unikraft]
UK_LIBS - The directory of all the external Unikraft libraries [default: $UK_WORKDIR/libs]
UK_APPS - The directory of all the template applications [default: $UK_WORKDIR/apps]
```

> After successfully running the above commands, we now have kraft installed on your system and we can get to building and running unikernels.

Author: ```suggestion
After successfully running the above commands, we now have kraft installed on our system and we can get to building and running unikernels.
```

#
[...]
CC      libkvmplat: trace.common.o
CC      libkvmplat: traps.isr.o
CC      libkvmplat: cpu_features.common.o
[...]
CC      libnolibc: errno.o
CC      libnolibc: stdio.o
CC      libnolibc: ctype.o
[...]
LD      hello_kvm-x86_64.ld.o
OBJCOPY hello_kvm-x86_64.o
LD      hello_kvm-x86_64.dbg
SCSTRIP hello_kvm-x86_64
GZ      hello_kvm-x86_64.gz
LN      hello_kvm-x86_64.dbg.gdb.py
[...]
Successfully built unikernels:

  => build/hello_kvm-x86_64
  => build/hello_kvm-x86_64.dbg (with symbols)

[...]
To instantiate, use: kraft run
[...]
Starting VM...
[...]
                   Tethys 0.5.0~b8be82b
Hello world!
Arguments:  "/home/razvan/hello/build/hello_kvm-x86_64" "console=ttyS0"
```

In the snippet above, we selected parts of the output showing what `kraft` does behind the scenes:

1. It downloads the `helloworld` application repository in the `hello/` directory.
1. It configures the repository, resulting in a `.config` file.
1. It builds the required components, resulting in the `build/hello_kvm-x86_64` unikernel image.
1. It runs the image, resulting in QEMU/KVM being run, with the "Hello world!" message getting printed.

All that magic is done using one command.

> A closer inspection of the `hello/` folder reveals it is a clone of the [app-helloworld repository](https://github.com/unikraft/app-helloworld) and it stores the resulting configuration file (`.config`) and resulting build folder (and images) (`build/):

Author: ```suggestion
A closer inspection of the `hello/` folder reveals it is a clone of the [app-helloworld repository](https://github.com/unikraft/app-helloworld) and it stores the resulting configuration file (`.config`) and resulting build folder (and images) (`build/`):
```

To run an already-built unikernel image, we use `kraft run`:
```
$ kraft run
[...]

                   Tethys 0.5.0~b8be82b
Hello world!
[...]
```

If we want to be more specific, we could use:
```
$ kraft run -p kvm -m x86_64
```

This command is useful in the case we have multiple images built (for differing platforms and architectures).
We can then select which one to run.

For example, we can use the commands below to configure, build and run a helloworld image for the `linuxu` platform.
```
kraft configure -p linuxu -m x86_64
kraft build
kraft run -p linuxu -m x86_64
```

You can now alter between running the `linuxu` and the `kvm` built images by using `kraft run` with the appropriate arguments.

### More on kraft

Of course, this is the most basic way you can use `kraft`, but there are many other options.
To see every option `kraft` has to offer, you can simply type `kraft -h`.
If you want to know about a certain command, just follow it with the `-h` option.
For example, if I wanted to know more about the configure command, I would type `kraft configure -h`.

### Manually Building the helloworld Application

Let's now learn how to build the app manually, without `kraft`.

> We won't go into much detail, this will be handled more thoroughly in session 02: Behind the Scenes.

Author: ```suggestion
We won't go into too much detail, this will be handled more thoroughly in session 02: Behind the Scenes.
```

```
$ make menuconfig
```

Looks like we are met with an error:
```
$ make menuconfig
Makefile:9: recipe for target 'menuconfig' failed
make: *** [menuconfig] Error 2
```

We look in the `Makefile`:
```
$ cat -n Makefile
     1  UK_ROOT ?= $(PWD)/../../unikraft
     2  UK_LIBS ?= $(PWD)/../../libs
     3  LIBS :=
     4
     5  all:
     6          @$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS)
     7
     8  $(MAKECMDGOALS):
     9          @$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) $(MAKECMDGOALS)
```
The underlying build / configuration system expects the Unikernel (`UK_ROOT`) to be located at `../../unikraft` from the current directory, which is very likely not the case.
Recall that the build system makes use of some important environment variables, namely `UK_WORKDIR`, `UK_ROOT` and `UK_LIBS`.
So, in order to properly inform the build system of our current location, we will have to manually set these by prefixing whatever build command we send with the hardcoded values of where our `Unikraft` work directory is.
```
$ UK_WORKDIR=~/.unikraft UK_ROOT=~/.unikraft/unikraft UK_LIBS=~/.unikraft/libs make menuconfig
```

> Note: This menu is also available through the `kraft menuconfig` command, which rids you of the hassle of manually setting the environment variables.

Author: ```suggestion
**Note**: This menu is also available through the `kraft menuconfig` command, which rids you of the hassle of manually setting the environment variables.
```

```

Next, we include it in our `Makefile`:
```
LIBS := $(UK_LIBS)/lwip:$(UK_LIBS)/newlib
```

And now, for the final step, through `make menuconfig` make sure you have selected `libnewlib` as well as `9pfs: 9p filesystem` inside the `Library Configuration` menu.
We will also check these options inside `Library Configuration` -> `vfscore: Configuration`:

![fs selection menu](/docs/sessions/01-baby-steps/images/menuconfig_select_fs.png)

![fs2 selection menu](/docs/sessions/01-baby-steps/images/menuconfig_select_fs2.png)

![fs3 selection menu](/docs/sessions/01-baby-steps/images/menuconfig_select_fs3.png)

What is more, you should also have present in the current directory an additional directory called `fs0`:
```
mkdir fs0
```
And so, `fs0` will contain whatever files you create, read from or write to from within your unikernel.

For now, just make sure it successfully builds. If it does, move on to the next work item.

## 04. Store Strings

For the final work item, you will have to update the source code from the second task, so that it stores in a file the received string before sending the encoded one back to the client.

> In order to achieve this, you must have the previous work item completed

Author: ```suggestion
In order to achieve this, you must have the previous work item completed.
```

 > 01-hello-world_linuxu-x86_64
   01-hello-world_kvm-x86_64
   01-hello-world_xen-x86_64
   01-hello-world_linuxu-arm64
   01-hello-world_kvm-arm64
   01-hello-world_xen-arm64
```
We have 6 options (2 arhitectures x 3 platforms).
Once we select one, the configuration will be updated.

The alternate way (non-interactive) is to pass arguments to `kraft configure` to select the desired platform and architecture.
For example, if we want to use x86_64 and KVM, we use:
```
$ kraft configure -p kvm -m x86_64
```

#### Build

Everything is set up now, all we have left to do is tell the build system to do its magic:
```
$ kraft build
[...]
Successfully built unikernels:

  => build/01-hello-world_kvm-x86_64
  => build/01-hello-world_kvm-x86_64.dbg (with symbols)

To instantiate, use: kraft run
```

This results in the creation of two unikernel image files:
1. `build/01-hello-world_kvm-x86_64` - the main image file
2. `build/01-hello-world_kvm-x86_64.dbg` - the image file with debug information (useful for debugging, duh!)

And that's it! Our final unikernel binary is ready to be launched from the `build/` directory.

#### Run

> To run an already-build unikernel image, we use `kraft run`:

Author: ```suggestion
To run an already-built unikernel image, we use `kraft run`:
```

        @$(MAKE) -C $(UK_ROOT) A=$(PWD) L=$(LIBS) $(MAKECMDGOALS)
```

#### Configure

Now, we configure it through `make menuconfig`.

![lwip selection menu](/docs/sessions/01-baby-steps/images/menuconfig_select_lwip.png)

![lwip2 selection menu](/docs/sessions/01-baby-steps/images/menuconfig_select_lwip2.png)

If you noticed, the menu also automatically selected some other internal components that would be required by `lwip`.
Now `Save` and `Exit` the configuration and run `make`.

#### Build

```

> $ make run

Author: ```suggestion
$ make
```

**Note**: We didn't go into configuring a functional network connection and actually querying the HTTP server.
This is a bit more complicated and is outside the scope of this session.

## Summary

`kraft`is an extremely useful tool for quickly deploying unikernel images.
It abstracts away many factors that would normally increase the difficulty of such tasks.
Through just a simple set of a few commands we can build and run a set of fast and secure unikernel images with low memory footprint.

## Practical Work

## 01. Echo-back Server

You will have to implement a simple echo-back server in C for the KVM platofrm.
The application will have to be able to open a socket on `172.44.0.2:1234` and send back to the client whatever the client sends to the server.
If the client closes the connection, the server will automatically close.

Enter the `work/01-echo-back/` directory.
Check the source code file (`main.c`) and support files.
Work on the contents to have a viable echo-back server implementation.
Things to consider:

* You will need some network client utility such as `netcat`.
* You will need the Lightweight TCP/IP stack library (lwip): https://github.com/unikraft/lib-lwip

> * You will have to update build and support files in the `work/01-echo-back/` directory.

Author: ```suggestion
* You will have to update the build and support files in the `work/01-echo-back/` directory.
```
